### PR TITLE
[ncp] add border routing InfraIf setup

### PR DESCRIPTION
**Pull request by Irving-cl**
This PR implements InfraIf SetUp on NCP.

See this [comment](https://github.com/openthread/ot-br-posix/issues/2398#issuecomment-2348518919) for the whole picture of this series of changes.

This PR also implements NCP based version of `otPlatInfraIfHasAddress`  by the way. The PR adds a unit test to test against the code in `NcpBase`.

### CODE REVIEW

File: src/lib/spinel/spinel.h [hunk 1]
```diff
@@ -4946,6 +4946,21 @@ enum
...
 
+    SPINEL_PROP_INFRA_IF__BEGIN = 0x910,
+
+    /// Infrastructure interface setup.
+    /** Format: `LbA(6)`
+     * Type: Write
+     *
+     * `L`: The infrastructure interface index.
+     * `b`: If the infrastrue interface is running.
+     * `A(6)`: The IPv6 addresses of the infrastructure interface.
```
Comment (abtink) [inline src/lib/spinel/spinel.h:4957]: The code parsing this property seems to decode it as array of structs where each struct contains an IPv6 address. Should we change this to `A(t(6))` to be match the decoding logic

Or we can consider changing the decoding logic so it treats it as `A(6)`?

**Review by abtink — COMMENTED** [review]
Looks good overall. Thanks @Irving-cl 
Couple of suggestion/questions below

File: src/ncp/ncp_base.hpp [hunk 2]
```diff
@@ -746,6 +757,20 @@ class NcpBase
...
+    {
+    public:
+        Ip6Address(void) { memset(mFields.m8, 0, sizeof(mFields.m8)); }
+
+        Ip6Address(const otIp6Address &aAddr) { memcpy(mFields.m8, &aAddr, sizeof(otIp6Address)); }
+    };
+
+    static constexpr uint8_t                kMaxInfraIfAddrs = 10;
+    ot::Array<Ip6Address, kMaxInfraIfAddrs> mInfraIfAddrs;
+    uint32_t                                mInfraIfIndex = 0;
```
[the remark this answers is not in the document]
Comment (Irving-cl) [reply src/ncp/ncp_base.hpp]: Make sense! Have implemented it without `ot::Array` and `Equatable`.

File: src/lib/spinel/spinel.h (hunk 1 above)
Comment (Irving-cl) [reply src/lib/spinel/spinel.h:4957]: Updated the decoding logic and treat it as `A(6)` since that's a little bit more efficient.

**Review by zhanglongxia — APPROVED** [review]
LGTM

File: src/ncp/ncp_base.hpp [hunk 3]
```diff
@@ -746,6 +755,16 @@ class NcpBase
...
     uint64_t mLogTimestampBase; // Timestamp base used for logging
 
+#if OPENTHREAD_FTD && OPENTHREAD_CONFIG_NCP_INFRA_IF_ENABLE
+    otError InfraIfAddAddress(const otIp6Address &aAddress);
+    bool    InfraIfContainsAddress(const otIp6Address &aAddress);
+
+    static constexpr uint8_t kMaxInfraIfAddrs = 10;
+    otIp6Address             mInfraIfAddrs[kMaxInfraIfAddrs];
+    uint8_t                  mInfraIfAddrCount = 0;
+    uint32_t                 mInfraIfIndex     = 0;
```
Comment (abtink) [inline src/ncp/ncp_base.hpp:765]: Could we initialize these in the `NcpBase` constructor instead?

We generally prefer initializing member variables within the constructor in the OT repository. This approach offers a couple of advantages:
- Consistency: It keeps the initialization logic centralized and easier to follow.
- Maintainability: All initial values are set in one place, simplifying debugging and future modifications.
- Performance: Some older/embedded toolchains might generate less efficient code for in-place member initialization.

**Review by abtink — APPROVED** [review]
Looks good. Thanks.
Couple of smaller suggestions below:

**Review by jwhui — APPROVED** [review]

**Pull request merged**